mixer/adapter/kubernetes2/cache.go

```diff
@@ -0,0 +1,266 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"errors"
+	"reflect"
+	"sync"
+	"time"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/mixer/pkg/adapter"
+)
+
+type (
+	// internal interface used to support testing
+	cacheController interface {
+		Run(<-chan struct{})
+		GetPod(string) (*v1.Pod, bool)
+		HasSynced() bool
+	}
+
+	controllerImpl struct {
+		clientset     kubernetes.Interface
+		env           adapter.Env
+		pods          cache.SharedInformer
+		mutationsChan chan resourceMutation
+
+		ipPodMap      map[string]string
+		ipPodMapMutex *sync.RWMutex
+	}
+
+	// used to send updates to the logger
+	resourceMutation struct {
+		kind eventType
+		obj  interface{}
+	}
+
+	eventType int
+)
+
+const (
+	addition eventType = iota
+	update
+	deletion
+)
+
+// mutationBufferSize sets the limit on how many mutation events can be
+// outstanding at any moment. 100 is chosen as a reasonable default to start.
+// TODO: make this configurable
+const mutationBufferSize = 100
+
+// errorDelay controls how long the logger waits when encountering an error
+// during logging. This should only ever happen when the underlying cache has
+// not yet synced (at which point we need to wait before doing any further
+// processing).
+// TODO: make this configurable
+const errorDelay = 1 * time.Second
+
+const debugVerbosityLevel = 4
+
+// Responsible for setting up the cacheController, based on the supplied client.
+// It configures the index informer to list/watch pods and send update events
+// to a mutations channel for processing (in this case, logging).
+func newCacheController(clientset *kubernetes.Clientset, refreshDuration time.Duration, env adapter.Env) cacheController {
+	c := &controllerImpl{
+		clientset:     clientset,
+		env:           env,
+		mutationsChan: make(chan resourceMutation, mutationBufferSize),
+		ipPodMapMutex: &sync.RWMutex{},
+		ipPodMap:      make(map[string]string),
+	}
+
+	namespace := "" // todo: address unparam linter issue
+
+	c.pods = cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
+				return clientset.Pods(namespace).List(opts)
+			},
+			WatchFunc: func(opts metav1.ListOptions) (watch.Interface, error) {
+				return clientset.Pods(namespace).Watch(opts)
+			},
+		},
+		&v1.Pod{},
+		refreshDuration,
+		cache.Indexers{},
```

Comment:
You can create the index of IP address to the pod here. This way, you don't have to keep `ipPodMap` anymore, and you don't have to have the event handler below.